Simplify YAML configuration

srp3rd opened the pull request.

srp3rd committed:
Changed YAML configuration to: specify brokers as a list, add "butler_repo"
Removed requirement to add "/" to the end of *_prefix strings.
Removed requirement of having "fs_prefix".  If it doesn't exist, it defaults to ""

srp3rd committed:
Additional YAML configuration file explanation added.
Added additional tests to tests/test_mapper.py

srp3rd committed:
Add quotes to keyword in README.md

srp3rd committed:
Fix Ruff complaint

timj commented on python/lsst/ctrl/ingestd/config.py:
```diff
@@ -38,24 +38,28 @@ def __init__(self, filename: str):
         with open(filename) as file:
```
Consider migrating to a pydantic model in the future. This will make a lot of this code much easier (and give you validation).

srp3rd replied:
Thank you for this suggestion.  I made a task to do that.

airnandez commented:
Thanks for these improvements. Here are my comments:

- I suggest to make explicit in the README file which items in the configuration file are required and which are optional. For the optionals (e.g. `num_messages`) mention what is the default value.
- according to the README, the example YAML file has Kafka topics `XRD1` and `XRD2` while in reality, the example has topics `XRD1-test` and `XRD2-test`
- the description of `butler_repo` could mention that there are several ways to specify the repo: via its location, via its `butler.yaml` configuration, via an alias present in file pointed to by `$DAF_BUTLER_REPOSITORY_INDEX`
- the description of the optional `timeout` is not included in the README
- are the default values for `num_messages` (1 message) and `timeout` (1 second) suitable for this daemon?
- a description of `rucio_prefix` would be welcome
- I suspect the log timestamps are displayed in the local time zone but include the suffix `Z` [here](https://github.com/lsst-dm/ctrl_ingestd/blob/af37e110aa08f5b297788eee154a1bc8468b7e0f/bin.src/ingestd#L30)
- using the hostname as Kafka client [here](https://github.com/lsst-dm/ctrl_ingestd/blob/af37e110aa08f5b297788eee154a1bc8468b7e0f/python/lsst/ctrl/ingestd/ingestd.py#L61) precludes running more than one `ingestd` instance for the same repo on the same host, in case that is needed to consume more messages
- I wonder if [these initializations](https://github.com/lsst-dm/ctrl_ingestd/blob/af37e110aa08f5b297788eee154a1bc8468b7e0f/python/lsst/ctrl/ingestd/mapper.py#L53) would better go in the constructor, since you are basically inspecting the contents of the configuration file and selecting appropriate defaults for _every_ message, while it could be done once and for all for each topic at initialization time of the `Mapper` object. The `rewrite()` method would do the actual string replace, which is the operation necessary when processing every incoming Kafka message

srp3rd commented:
> Thanks for these improvements. Here are my comments:
> 
> * I suggest to make explicit in the README file which items in the configuration file are required and which are optional. For the optionals (e.g. `num_messages`) mention what is the default value.

Added

> * according to the README, the example YAML file has Kafka topics `XRD1` and `XRD2` while in reality, the example has topics `XRD1-test` and `XRD2-test`

Fixed

> * the description of `butler_repo` could mention that there are several ways to specify the repo: via its location, via its `butler.yaml` configuration, via an alias present in file pointed to by `$DAF_BUTLER_REPOSITORY_INDEX`

Added

> * the description of the optional `timeout` is not included in the README

Added

> * are the default values for `num_messages` (1 message) and `timeout` (1 second) suitable for this daemon?

The default for `num_messages` is expected to be changed to allow larger batches of messages to be processed.  I couldn't really decide what would be appropriate, since I thought whatever I picked probably wouldn't what people would use as a default anyway.  Having said that, I set the default to 50 messages.   The default for `timeout` of 1 second is what I believe is appropriate.

> * a description of `rucio_prefix` would be welcome

Added

> * I suspect the log timestamps are displayed in the local time zone but include the suffix `Z` [here](https://github.com/lsst-dm/ctrl_ingestd/blob/af37e110aa08f5b297788eee154a1bc8468b7e0f/bin.src/ingestd#L30)

I grabbed that string from another daemon I run, where all the machines are synced to UTC.  I'll remove that `Z`.

> * using the hostname as Kafka client [here](https://github.com/lsst-dm/ctrl_ingestd/blob/af37e110aa08f5b297788eee154a1bc8468b7e0f/python/lsst/ctrl/ingestd/ingestd.py#L61) precludes running more than one `ingestd` instance for the same repo on the same host, in case that is needed to consume more messages

I've added an optional "client_id", which defaults to the hostname, but can be specified if you wish to do what you say above.  This has also been added to the README.md file.

> * I wonder if [these initializations](https://github.com/lsst-dm/ctrl_ingestd/blob/af37e110aa08f5b297788eee154a1bc8468b7e0f/python/lsst/ctrl/ingestd/mapper.py#L53) would better go in the constructor, since you are basically inspecting the contents of the configuration file and selecting appropriate defaults for _every_ message, while it could be done once and for all for each topic at initialization time of the `Mapper` object. The `rewrite()` method would do the actual string replace, which is the operation necessary when processing every incoming Kafka message

I do see what you're saying.  I believe it's more appropriate for the Config object to check and set the `endswith` values for `rse_prefix` and `fs_prefix`, rather than in Mapper, so I'm going to move it there.

srp3rd committed:
Update README.md file
Remove errand "Z" from logging string
Move rse_prefix and fs_prefix end of string checks to Config, where
they can be dealt with once on initialization
Added new optional client_id to configuration file, which it is set to hostname by default

srp3rd committed:
Fix RUFF complaints

srp3rd committed:
Fix another ruff complaint

srp3rd committed:
Add a few more unit tests to test config

srp3rd committed:
Fixed ruff error
